### PR TITLE
Exception Handling Section

### LINKED ISSUE #49: Consider providing guidance on exception handling
I used this guide to refresh my JAX-RS knowledge. Unfortunately, when I went to apply that, I was getting a 500 error back in my client but there was nothing useful in the server logs (even with debug tracing).

It turned out to be a dumb mistake on my part (ArrayOutOfBoundsException), but is there a suggested way for configuring a root exception handler so that future mistakes like that would at least get logged?

I ended up following the advice at https://stackoverflow.com/a/19680115/161022 to register an ExceptionMapper and inserted a log statement there. If that is the recommended approach, it might be nice to add something like that to this guide so that anyone starting from here will have unhandled exceptions logged for them.  Let me know if I'm missing something!

### PR DESCRIPTION
**Pull request by AustinSeto**
Create section in guide to handle exceptions

Addresses https://github.com/OpenLiberty/guide-rest-intro/issues/49

### CODE REVIEW

**Review by griffinhadfield — CHANGES_REQUESTED** [review]
- [x] `src/main/java/io/openliberty/guides/rest/exceptions` should be created in the start directory with a `.gitkeep` in it. 
- [x] Add file hotspots for `GenericExceptionMapper.java` and `NotFoundExceptionMapper.java`
- [x] Consider adding hotspots for `toResponse()` and `ErrorResponse` in the excerpt from the Exception Handling section below:
 > When JAX-RS sees any Throwable, it will search for a matching ExceptionMapper interface that has been marked with the @Provider annotation. It will pass the Throwable to the toResponse() method to construct a response using the ErrorResponse class.

**Comment by AustinSeto** [comment]
@griffinhadfield changes made, hotspots added

**Review by griffinhadfield — APPROVED** [review]
Feedback has been addressed, LGTM.

File: README.adoc [hunk 1]
```diff
@@ -198,6 +198,68 @@ include::{common-includes}/devmode-build.adoc[]
...
 
+// =================================================================================================
+// Exception Handling
+// =================================================================================================
+
+== Exception Handling
+
+If a resource doesn't exist or the service experiences an exception, the client will receive either an empty response or an HTML error page respectively.
+A client should not receive either of these.
+Clients should receive JSON responses.
```
Comment (andymc12) [inline README.adoc]: I would rephrase this section.  I'm not sure that the client will always receive an empty response or an HTML error page.  And ideally, the client would receive the error response in the same MediaType (MIME type) that they would receive a normal response in (i.e. it could be text, xml, etc.).

Maybe something like this:

Most application developers will want to ensure that when an exception occurs in the application (such as the client requesting a resource that does not exist or posting an invalid resource, etc.) that it is handled consistently and in a way that allows the client to fully understand the failure (i.e. bad input, server outage, etc.). This can be accomplished using the exception handling mechanisms in JAX-RS such as sending a `Response` via a `WebApplicationException` (or subclass) or by using `ExceptionMapper` providers.

File: README.adoc [hunk 2]
```diff
@@ -198,6 +198,68 @@ include::{common-includes}/devmode-build.adoc[]
...
+// Exception Handling
+// =================================================================================================
+
+== Exception Handling
+
+If a resource doesn't exist or the service experiences an exception, the client will receive either an empty response or an HTML error page respectively.
+A client should not receive either of these.
+Clients should receive JSON responses. 
+
+This will be handled using an `ExceptionMapper` which maps exceptions to responses and builds appropriate JSON payloads.
```
Comment (andymc12) [inline README.adoc]:
```suggestion
The following example uses an `ExceptionMapper` which maps exceptions to responses and builds appropriate JSON payloads. 
```

File: README.adoc [hunk 3]
```diff
@@ -198,6 +198,68 @@ include::{common-includes}/devmode-build.adoc[]
...
+
+The [hotspot file=1]`GenericExceptionMapper` class will be used for any `Throwable` that does not have a more specific `ExceptionMapper` interface implemented. 
+
+[role="code_command hotspot file=2", subs="quotes"]
+----
+#Create the `NotFoundExceptionMapper` class.#
+`src/main/java/io/openliberty/guides/rest/exceptions/NotFoundExceptionMapper.java`
+----
+
+NotFoundExceptionMapper.java
```
Comment (andymc12) [inline README.adoc]: We discussed in slack that it might be better to use a different exception type here - ideally an application-specific, business exception (like `WidgetNotFoundException`) since `NotFoundException` is a subclass of `WebApplicationException` - and it could hold it's own `Response` object (and thus doesn't need to be mapped).

**Review by andymc12 — CHANGES_REQUESTED** [review]
Suggested (via slack) describing how to set a `Response` in a `WebApplicationException` (or subclass) - and minor changes in wording.

File: README.adoc [hunk 4]
```diff
@@ -198,6 +201,67 @@ include::{common-includes}/devmode-build.adoc[]
...
+
+The [hotspot file=1]`GenericExceptionMapper` class will be used for any `Throwable` that does not have a more specific `ExceptionMapper` interface implemented. 
+
+[role="code_command hotspot file=2", subs="quotes"]
+----
+#Create the `PropertyNotFoundExceptionMapper` class.#
+`src/main/java/io/openliberty/guides/rest/exceptions/NotFoundExceptionMapper.java`
+----
+
+NotFoundExceptionMapper.java
```
Comment (andymc12) [inline README.adoc]:
```suggestion
PropertyNotFoundExceptionMapper.java
```

File: README.adoc [hunk 5]
```diff
@@ -198,6 +201,67 @@ include::{common-includes}/devmode-build.adoc[]
...
+
+When JAX-RS sees any `Throwable`, it will search for a matching `ExceptionMapper` interface that has been marked with the [hotspot=provider file=1]`@Provider` annotation. 
+It will pass the `Throwable` to the [hotspot=toResponse file=1]`toResponse()` method to construct a response using the [hotspot file=0]`ErrorResponse` class.
+
+The [hotspot file=1]`GenericExceptionMapper` class will be used for any `Throwable` that does not have a more specific `ExceptionMapper` interface implemented. 
+
+[role="code_command hotspot file=2", subs="quotes"]
+----
+#Create the `PropertyNotFoundExceptionMapper` class.#
+`src/main/java/io/openliberty/guides/rest/exceptions/NotFoundExceptionMapper.java`
```
Comment (andymc12) [inline README.adoc]:
```suggestion
`src/main/java/io/openliberty/guides/rest/exceptions/PropertyNotFoundExceptionMapper.java`
```

**Review by andymc12 — COMMENTED** [review]
Just a couple tweaks - otherwise, looks good.  Thanks!

**Review by andymc12 — APPROVED** [review]
Looks good!  Thanks @AustinSeto!

**Comment by gkwan-ibm** [comment]
close
- need to re-investigate the issue
